src/Analysis/Ast/Impl/Analyzer/Evaluation/ExpressionEval.Generics.cs

```diff
@@ -59,40 +63,62 @@ internal sealed partial class ExpressionEval {
-            // See which ones are generic parameters as defined by TypeVar() 
-            // and which are specific types. Normally there should not be a mix.
-            var genericTypeArgs = indices.OfType<IGenericTypeDefinition>().ToArray();
-            var specificTypes = indices.Where(i => !(i is IGenericTypeDefinition)).OfType<IPythonType>().ToArray();
-
-            if (genericTypeArgs.Length > 0 && genericTypeArgs.Length != indices.Count) {
-                // TODO: report that some type arguments are not declared with TypeVar.
+        private bool GenericClassParameterValid(IGenericTypeDefinition[] genericTypeArgs, IReadOnlyList<IMember> indices, Expression expr) {
+            if (genericTypeArgs.Length == 0) {
+                return false;
             }
-            if (specificTypes.Length > 0 && specificTypes.Length != indices.Count) {
-                // TODO: report that arguments are not specific types or are not declared.
+
+            // All arguments to Generic must be type parameters
+            // e.g. Generic[T, str] throws a runtime error
+            if (genericTypeArgs.Length != indices.Count) {
+                ReportDiagnostics(Module.Uri, new DiagnosticsEntry(
+                    Resources.GenericArgumentsNotAllTypeParameters,
+                    GetLocation(expr).Span,
+                    ErrorCodes.GenericArguments,
+                    Severity.Error,
+                    DiagnosticSource.Analysis));
+                return false;
+            }
+
+            // All arguments to Generic must be distinct
+            if (genericTypeArgs.Distinct().Count() != genericTypeArgs.Length) {
+                ReportDiagnostics(Module.Uri, new DiagnosticsEntry(
+                   Resources.GenericArgumentsNotAllUnique,
+                   GetLocation(expr).Span,
+                   ErrorCodes.GenericArguments,
+                   Severity.Error,
+                   DiagnosticSource.Analysis));
+                return false;
             }
 
+            return true;
+        }
+
+        /// <summary>
+        /// Given generic type and list of arguments in the expression like
+        /// Generic[T1, T2, ...] or List[str] creates generic class base
+        /// (if the former) on specific type (if the latter).
+        /// </summary>
+        private IMember CreateSpecificTypeFromIndex(IGenericType gt, IReadOnlyList<IMember> args, Expression expr) {
+            var genericTypeArgs = args.OfType<IGenericTypeDefinition>().ToArray();
+
             if (gt.Name.EqualsOrdinal("Generic")) {
-                // Generic[T1, T2, ...] expression. Create generic base for the class.
-                if (genericTypeArgs.Length > 0) {
-                    return new GenericClassParameter(genericTypeArgs, Module);
-                } else {
-                    // TODO: report too few type arguments for Generic[].
```

Comment:
When there are no arguments to Generic[] then there is a syntax error.

Reply:
I've sent you an email. This one is a separate item. ex `class A(Generic[]):`

Reply:
I said (and we observed) that it was a syntax error and probably didn't need to be handled, but during the analysis I think we'd actually have partial results and end up with an empty indexing (which we would want in order to provide completion), so my impression was wrong here.

Reply:
So long as we don't bail out of the generic code due to the missing expression, we shouldn't send a diagnostic for this. We looked at how the parser behaves, and the index expression is an error expression (and would have already had a diagnostic earlier). Doing it again probably doesn't provide any value.